src/workerd/jsg/modules-new.c++

```diff
@@ -706,76 +714,80 @@ v8::MaybeLocal<v8::Module> resolveCallback(v8::Local<v8::Context> context,
-        // Make sure that percent-encoding in the path is normalized so we can match correctly.
-        auto normalized = url.clone(Url::EquivalenceOption::NORMALIZE_PATH);
-        ResolveContext resolveContext = {
-          .type = type,
-          .source = ResolveContext::Source::STATIC_IMPORT,
-          .specifier = normalized,
-          .referrer = referrerUrl,
-          .rawSpecifier = spec.asPtr(),
-        };
-        // TODO(soon): Add import assertions to the context.
+        case Module::Type::BUILTIN: {
+          type = ResolveContext::Type::BUILTIN;
+          break;
+        }
+        case Module::Type::BUILTIN_ONLY: {
+          type = ResolveContext::Type::BUILTIN_ONLY;
+          break;
+        }
+        case Module::Type::FALLBACK: {
+          type = ResolveContext::Type::BUNDLE;
+          break;
+        }
+      }
+      return entry.specifier.specifier;
+    }).orDefault(ModuleBundle::BundleBuilder::BASE);
 
-        return registry.resolve(js, resolveContext);
+    // If Node.js Compat v2 mode is enable, we have to check to see if the specifier
+    // is a bare node specifier and resolve it to a full node: URL.
+    if (isNodeJsCompatEnabled(js)) {
+      KJ_IF_SOME(nodeSpec, checkNodeSpecifier(spec)) {
+        spec = kj::mv(nodeSpec);
       }
+    }
 
-      js.throwException(js.error(kj::str("Invalid module specifier: "_kj, specifier)));
-    }, [&](Value exception) -> v8::MaybeLocal<v8::Module> {
-      // If there are any synchronously thrown exceptions, we want to catch them
-      // here and convert them into a rejected promise. The only exception are
-      // fatal cases where the isolate is terminating which won't make it here
-      // anyway.
-      js.v8Isolate->ThrowException(exception.getHandle(js));
-      return v8::MaybeLocal<v8::Module>();
-    });
-  } catch (...) {
-    kj::throwFatalException(kj::getCaughtExceptionAsKj());
-  }
+    KJ_IF_SOME(url, referrerUrl.tryResolve(spec)) {
+      // Make sure that percent-encoding in the path is normalized so we can match correctly.
+      auto normalized = url.clone(Url::EquivalenceOption::NORMALIZE_PATH);
+      ResolveContext resolveContext = {
+        .type = type,
+        .source = ResolveContext::Source::STATIC_IMPORT,
+        .specifier = normalized,
+        .referrer = referrerUrl,
+        .rawSpecifier = spec.asPtr(),
+      };
+      // TODO(soon): Add import assertions to the context.
+
+      return registry.resolve(js, resolveContext);
+    }
+
+    js.throwException(js.error(kj::str("Invalid module specifier: "_kj, specifier)));
+  }, [&](Value exception) -> v8::MaybeLocal<v8::Module> {
+    // If there are any synchronously thrown exceptions, we want to catch them
+    // here and convert them into a rejected promise. The only exception are
+    // fatal cases where the isolate is terminating which won't make it here
+    // anyway.
+    js.v8Isolate->ThrowException(exception.getHandle(js));
```

Comment:
If we are only going to throw what we caught, we might use something like HandleScope and check similar to what we do in Node.js?

Reply:
Possibly. I'm going to be revisiting this specifically as I don't think the errors are being correctly handled right now. Will verify it with a couple of tests in the next iteration.